pika/adapters/select_connection.py

```diff
@@ -238,51 +234,38 @@ def poll(self, write_only=False):
             the adapter can write.
 
         """
-        # Build our values to pass into select
-        input_fileno, output_fileno, error_fileno = [], [], []
-
-        if self.events & READ:
-            input_fileno = [self.fileno]
-        if self.events & WRITE:
-            output_fileno = [self.fileno]
-        if self.events & ERROR:
-            error_fileno = [self.fileno]
 
         # Wait on select to let us know what's up
         try:
-            read, write, error = select.select(input_fileno,
-                                               output_fileno,
-                                               error_fileno,
-                                               SelectPoller.TIMEOUT)
+            read, write, error = select.select(
+                self.r_fds, self.w_fds, self.e_fds, self.get_next_timeout())
         except select.error as error:
-            return self._handler(self.fileno, ERROR, error)
+            if error.errno != errno.EINTR:
+                raise
 
         # Build our events bit mask
-        events = 0
-        if read:
-            events |= READ
-        if write:
-            events |= WRITE
-        if error:
-            events |= ERROR
 
-        if events:
-            self._handler(self.fileno, events, write_only=write_only)
+        for fileno in self.filenos:
+            events = 0
+            if fileno in read:
+                events |= READ
+            if fileno in write:
+                events |= WRITE
+            if fileno in error:
+                events |= ERROR
+           
+            if events:
+                handler = self.filenos[fileno]
```

Comment:
@wjps, it's conceivable that an earlier handler callback in this loop's cycle might call something that removes one of the the other handlers, so this code will break. `for fileno in self.filenos` might break, too, if `self.filenos` were modified during iteration. Same applies to the other pollers in this module. I think that you need to deal with it as follows, just to be on the safe side:

1. before the loop, save a copy: `filenos = self.filenos[:]`, then loop over the copy `for fileno in filenos:` instead of over `self.filenos`
2. Skip if `fileno not in self.filenos`

The problem might be more tricky if filenos are not sockets, but actual integer file descriptors, in which case closing one and opening another might re-use the same file descriptor number, so the check `fileno not in self.filenos` might not be enough.

Another complication might occur if an earlier handler callback (in the same loop cycle) modifies the events of interest for another fileno that was already registered in `self.filenos`, and the event that triggered is no longer of interest.